chain/chain/src/flat_storage_resharder.rs

```diff
@@ -1514,4 +1799,200 @@ mod tests {
         );
         assert_eq!(flat_store.get(right_child_shard, &buffered_receipt_key), Ok(None));
     }
+
+    /// Base test scenario for testing children catchup.
+    fn children_catchup_base(with_restart: bool) {
+        init_test_logger();
+        let (mut chain, mut resharder, scheduler) =
+            create_chain_resharder_scheduler::<DelayedScheduler>(simple_shard_layout());
+        let new_shard_layout = shard_layout_after_split();
+        let resharding_event_type = event_type_from_chain_and_layout(&chain, &new_shard_layout);
+        let ReshardingSplitShardParams {
+            parent_shard,
+            left_child_shard,
+            right_child_shard,
+            block_hash,
+            ..
+        } = match resharding_event_type.clone() {
+            ReshardingEventType::SplitShard(params) => params,
+        };
+        let manager = chain.runtime_adapter.get_flat_storage_manager();
+
+        // Do resharding.
+        assert!(resharder.start_resharding(resharding_event_type, &new_shard_layout).is_ok());
+
+        // Trigger the task to perform the parent split.
+        scheduler.call_split_shard_task();
+
+        // Simulate the chain going forward by seven blocks.
+        // Note that the last two blocks won't be yet 'final'.
+        const NUM_BLOCKS: u64 = 5;
+        let signer = Arc::new(create_test_signer("aa"));
+        for height in 1..NUM_BLOCKS + 1 {
+            let prev_block = chain.get_block_by_height(height - 1).unwrap();
+            let block = TestBlockBuilder::new(Clock::real(), &prev_block, signer.clone())
+                .height(height)
+                .build();
+            chain.process_block_test(&None, block).unwrap();
+        }
+        assert_eq!(chain.head().unwrap().height, NUM_BLOCKS);
+
+        // Manually add deltas into the children shards.
+        // For simplicity, add one new account at every height.
+        for height in 1..NUM_BLOCKS + 1 {
+            let prev_hash = *chain.get_block_by_height(height).unwrap().header().prev_hash();
+            let block_hash = *chain.get_block_by_height(height).unwrap().hash();
+
+            let new_account_left_child = account!(format!("oo{}", height));
+            let state_changes_left_child = vec![RawStateChangesWithTrieKey {
+                trie_key: TrieKey::Account { account_id: new_account_left_child.clone() },
+                changes: vec![RawStateChange {
+                    cause: StateChangeCause::InitialState,
+                    data: Some(new_account_left_child.as_bytes().to_vec()),
+                }],
+            }];
+            manager
+                .save_flat_state_changes(
+                    block_hash,
+                    prev_hash,
+                    height,
+                    left_child_shard,
+                    &state_changes_left_child,
+                )
+                .unwrap()
+                .commit()
+                .unwrap();
```

Comment:
nit: This is copy pasted, can you refactor to a method?

Reply:
Done!